src/lapic/lapic_msr.rs

```diff
@@ -6,70 +6,102 @@ use core::ops::Range;
 use paste;
 use x86_64::registers::model_specific::Msr;
 
+use crate::lapic::LocalApicMode;
+
+
+#[derive(Debug)]
+pub struct LocalApicRegister { msr: Msr, mmio_offset: u32 }
+
+impl LocalApicRegister {
+    pub fn new(register: (u32, u32)) -> Self {
+        Self { msr: Msr::new(register.0), mmio_offset: register.1 }
+    }
+
+    pub unsafe fn mmio_address(&self) -> u64 {
+        let base = Msr::new(IA32_APIC_BASE).read() & 0xFFFFFF000;
+        base + self.mmio_offset as u64
+    }
```

Comment:
This may be conceptually better served by an enum. What do you think? Also, it'd be nice if it could store the absolute MMIO address instead of just the offset. This way we can avoid passing in `mode` and reading the IA32_APIC_BASE MSR on every read and write. 

Maybe something like:

```rust
pub enum LocalApicMode {
    XApic { mmio_base: u64 },
    X2Apic,
}

pub enum LocalApicRegister {
    Mmio(u64),
    Msr(Msr),
}

impl LocalApicRegister {
    pub fn new(mode: LocalApicMode, loc: (u32, u32)) -> Self {
        match mode {
            XApic { mmio_base } -> Self::Mmio(mmio_base + loc.1 as u64),
            X2Apic -> Self::Msr(Msr::new(loc.0)),
        }
    }

    pub fn read(&self) -> u64 {
        match Self {
            Mmio(addr) -> ptr::read_volatile(addr as *const u64),
            Msr(msr) -> msr.read(),
        }
    }

    ...
}

...

impl LocalApicRegisters {
    pub fn new(mode: LocalApicMode) -> Self {
        Self {
            base: ...,
            id: LocalApicRegister::new(mode, ID),
            ...
        }
    }
}
```